spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScan.java

```diff
@@ -145,10 +145,12 @@ protected Statistics estimateStatistics(Snapshot snapshot) {
     }
 
     long numRows = 0L;
+    long defaultRowSizeInBytes = readSchema().defaultSize();
 
     for (CombinedScanTask task : tasks()) {
       for (FileScanTask file : task.files()) {
-        numRows += file.file().recordCount();
+        // TODO: if possible, take deletes also into consideration.
+        numRows += Math.min(file.length() / defaultRowSizeInBytes, file.file().recordCount());
```

Comment:
We always want to base the estimate on the number of records in the file and not on the file length. Calculations based on the file length are quite bad in practice because Parquet compression varies.

If you want to find out how many records from the file will actually be scanned by this task, then you could do something different, like finding the proportion of the file that is covered by the task and multiplying by the record count.